src/libraries/System.Private.CoreLib/src/System/Number.NumberToFloatingPointBits.cs

```diff
@@ -326,16 +1015,48 @@ private static ulong DigitsToUInt64(byte* p, int count)
             Debug.Assert((1 <= count) && (count <= 19));
 
             byte* end = (p + count);
-            ulong res = (ulong)(p[0] - '0');
+            ulong res = 0;
+
+            // parse batches of 8 digits with SWAR
+            while (end - p >= 8)
+            {
+                res = (res * 100000000) + ParseEightDigitsUnrolled(p);
+                p += 8;
+            }
 
-            for (p++; p < end; p++)
+            while (p != end)
             {
                 res = (10 * res) + p[0] - '0';
+                ++p;
             }
 
             return res;
         }
 
+        /// <summary>
+        /// Parse eight consecutive digits using SWAR
+        /// https://lemire.me/blog/2022/01/21/swar-explained-parsing-eight-digits/
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static uint ParseEightDigitsUnrolled(byte* chars)
+        {
+            // let's take the following value (byte*) 12345678 and read it unaligned :
+            // we get a ulong value of 0x3837363534333231
+            // 1. Subtract character '0' 0x30 for each byte to get 0x0807060504030201
+            // 2. Consider this sequence as bytes sequence : b8b7b6b5b4b3b2b1
+            // we need to transform it to b1b2b3b4b5b6b7b8 by computing :
+            // 10000 * (100 * (10*b1+b2) + 10*b3+b4) + 100*(10*b5+b6) + 10*b7+b8
+            // this is achieved by masking and shifting values
```

Comment:
This seems like an expensive way to do `System.Buffers.Binary.BinaryPrimitives.ReverseEndianness(x)` (which is a single instruction on most platforms).

Reply:
Hmmm, I think part of it is the comment here is a bit confusing.

* This is taking an ASCII string "12345678" and reading it to get `0x3837363534333231`
* We then want to normalize that to the base numbers (subtract `0x3030303030303030`) to get `0x0807060504030201`
* We want to reverse endianness to get `0x0102030405060708`
  * This is problematic for big endian machines which Mono may support
* We then want to pack the results down so we get `12345678` as an integer that fits in `uint`

Reply:
I think, particularly since big endian machines are a potential concern; that we either need to:
A. Use `BinaryPrimitives.ReadLittleEndian` here to get the `val`
B. Modify the algorithm to use `BinaryPrimitives.ReadBigEndian` and to do the packing step differently.

Reply:
@tannergooding The spirit here is that most systems (almost all of them) are little endian. So it is worth the effort to have a fast path for the little endian case, but it is much less well motivated to be concerned with the big endian case.

What we do in the C++ library is this...
```C++
fastfloat_really_inline uint64_t read_u64(const char *chars) {
  uint64_t val;
  ::memcpy(&val, chars, sizeof(uint64_t));
#if FASTFLOAT_IS_BIG_ENDIAN == 1
  // Need to read as-if the number was in little-endian order.
  val = byteswap(val);
#endif
  return val;
}
```
https://github.com/fastfloat/fast_float/blob/6d768256f13c927216f22995ab981c60f30c082a/include/fast_float/ascii_number.h#L28-L36

I think that's going to be equivalent to your `BinaryPrimitives.ReadLittleEndian` proposal.

Reply:
Right. `ReadLittleEndian` will be a `nop` on little endian systems and will insert the relevant 1 instruction `bswap` on other platforms.

It (`BigEndian`) is something that Mono supports, however, and so we need to support it here as well.